google/cloud/forseti/services/scanner/dao.py

```diff
@@ -18,150 +18,259 @@
+    violation_type = Column(String(256), nullable=False)
+
+    def __repr__(self):
+        """String representation.
+
+        Returns:
+            str: string representation of the Violation row entry.
+        """
+        string = ('<Violation(violation_type={}, resource_type={} '
+                  'rule_name={})>')
+        return string.format(
+            self.violation_type, self.resource_type, self.rule_name)
+
+class ViolationAccess(object):
+    """Facade for violations, implement APIs against violations table."""
+
+    def __init__(self, session):
+        """Constructor for the Violation Access.
+
+        Args:
+            session (Session): SQLAlchemy session object.
+        """
+        self.session = session
+
+    def create(self, violations, scanner_index_id):
+        """Save violations to the db table.
+
+        Args:
+            violations (list): A list of violations.
+            scanner_index_id (str): id of the `ScannerIndex` row for this
+                scanner run.
+        """
+        created_at_datetime = date_time.get_utc_now_datetime()
+        for violation in violations:
+            violation_hash = _create_violation_hash(
+                violation.get('full_name', ''),
+                violation.get('resource_data', ''),
+                violation.get('violation_data', ''),
+            )
+
+            violation = Violation(
+                created_at_datetime=created_at_datetime,
+                full_name=violation.get('full_name'),
+                resource_data=violation.get('resource_data'),
+                resource_id=violation.get('resource_id'),
+                resource_type=violation.get('resource_type'),
+                rule_index=violation.get('rule_index'),
+                rule_name=violation.get('rule_name'),
+                scanner_index_id=scanner_index_id,
+                violation_data=json.dumps(
+                    violation.get('violation_data')),
+                violation_hash=violation_hash,
+                violation_type=violation.get('violation_type')
+            )
+            self.session.add(violation)
+
+    def list(self, inv_index_id=None, scanner_index_id=None):
+        """List all violations from the db table.
+
+        Args:
+            inv_index_id (str): Id of the inventory index.
+            scanner_index_id (str): Id of the scanner index.
+
+        Returns:
+            list: List of Violation row entry objects.
+        """
+        if inv_index_id:
+            results = (
+                self.session.query(Violation, ScannerIndex)
+                .filter(and_(or_(
+                    ScannerIndex.scanner_status == 'SUCCESS',
+                    ScannerIndex.scanner_status == 'PARTIAL_SUCCESS'),
+                    ScannerIndex.inventory_index_id == inv_index_id))
+                .filter(Violation.scanner_index_id == ScannerIndex.id)
+                .all())
+            return [v for (v, _) in results]
```

Comment:
We have decided last week to avoid using list comprehensions because our IDE debuggers can not step through them, which becomes a support issue.  So, can you please write this out?
```
violations = []
for v, _ in results:
  violations.append(v)
return violations
```